org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/SnippetCompletionProposal.java

```diff
@@ -221,35 +238,60 @@ String getRecommendedLineSeprator() throws JavaModelException {
 			return Collections.emptyList();
 		}
 
-		for (Template template : templates) {
-			if (!javaContext.canEvaluate(template)) {
-				continue;
-			}
-			TemplateBuffer buffer = null;
-			try {
-				buffer = javaContext.evaluate(template);
-			} catch (BadLocationException | TemplateException e) {
-				JavaLanguageServerPlugin.logException(e.getMessage(), e);
-				continue;
-			}
-			if (buffer == null) {
-				continue;
-			}
-			String content = buffer.getString();
-			if (Strings.containsOnlyWhitespaces(content)) {
-				continue;
-			}
+		String uri = JDTUtils.toURI(cu);
+		Template[] availableTemplates = Arrays.stream(templates).filter(t -> javaContext.canEvaluate(t)).toArray(Template[]::new);
+		List<CompletionProposal> proposals = new ArrayList<>();
+		for (int i = 0; i < availableTemplates.length; i++) {
+			Template template = availableTemplates[i];
 			final CompletionItem item = new CompletionItem();
 			item.setLabel(template.getName());
-			item.setInsertText(content);
+			item.setKind(CompletionItemKind.Snippet);
+			item.setInsertTextFormat(InsertTextFormat.Snippet);
+			item.setInsertText(SnippetUtils.templateToSnippet(template.getPattern()));
 			item.setDetail(template.getDescription());
-			setFields(item, cu);
+
+			Map<String, String> data = new HashMap<>(3);
+			data.put(CompletionResolveHandler.DATA_FIELD_URI, uri);
+			data.put(CompletionResolveHandler.DATA_FIELD_REQUEST_ID, String.valueOf(response.getId()));
+			data.put(CompletionResolveHandler.DATA_FIELD_PROPOSAL_ID, String.valueOf(i));
```

Comment:
@rgrunber I just remember the reason why I need `i` is here. Index is kept in `data` field, so that in "resolve" stage corresponding proposal can be found.

Reply:
Right. In that case, I agree as well. The correct ordering is required, so being explicit about it as a good thing to do.